rust/lance-io/src/object_store/providers.rs

```diff
@@ -23,70 +27,278 @@ pub trait ObjectStoreProvider: std::fmt::Debug + Sync + Send {
+    ///
+    /// Calling this will also clean up any weak references to object stores that
+    /// are no longer valid.
+    pub fn active_stores(&self) -> Vec<Arc<ObjectStore>> {
+        let mut found_inactive = false;
+        let output = self
+            .active_stores
+            .read()
+            .expect("ObjectStoreRegistry lock poisoned")
+            .values()
+            .filter_map(|weak| match weak.upgrade() {
+                Some(store) => Some(store),
+                None => {
+                    found_inactive = true;
+                    None
+                }
+            })
+            .collect();
+
+        if found_inactive {
+            // Clean up the cache by removing any weak references that are no longer valid
+            let mut cache_lock = self
+                .active_stores
+                .write()
+                .expect("ObjectStoreRegistry lock poisoned");
+            cache_lock.retain(|_, weak| weak.upgrade().is_some());
+        }
+        output
+    }
+
+    /// Get an object store for a given base path and parameters.
+    ///
+    /// If the object store is already in use, it will return a strong reference
+    /// to the object store. If the object store is not in use, it will create a
+    /// new object store and return a strong reference to it.
     pub async fn get_store(
         &self,
         base_path: Url,
         params: &ObjectStoreParams,
-    ) -> Result<ObjectStore> {
-        // TODO: caching
+    ) -> Result<Arc<ObjectStore>> {
+        let cache_path = cache_url(&base_path);
+        let cache_key = (cache_path, params.clone());
+
+        // Check if we have a cached store for this base path and params
+        {
+            let maybe_store = self
+                .active_stores
+                .read()
+                .expect("ObjectStoreRegistry lock poisoned")
+                .get(&cache_key)
+                .cloned();
+            if let Some(store) = maybe_store {
+                if let Some(store) = store.upgrade() {
+                    return Ok(store);
+                } else {
+                    // Remove the weak reference if it is no longer valid
+                    let mut cache_lock = self
+                        .active_stores
+                        .write()
+                        .expect("ObjectStoreRegistry lock poisoned");
+                    if let Some(store) = cache_lock.get(&cache_key) {
+                        if store.upgrade().is_none() {
+                            // Remove the weak reference if it is no longer valid
+                            cache_lock.remove(&cache_key);
+                        }
+                    }
+                }
+            }
+        }
+
         let scheme = base_path.scheme();
         let provider = self.get_provider(scheme).ok_or_else(|| {
+            let valid_schemes = self
+                .providers
+                .read()
+                .expect("ObjectStoreRegistry lock poisoned")
```

Comment:
This `expect` message isn't unique.  If this triggers we won't know which statement triggers it.  Can we do something like `LanceOptionExt::expect_ok` for results?  That adds the location and it avoids needed to specify a string.

Reply:
Done.